packages/devtools_app/lib/src/charts/treemap.dart

```diff
@@ -1,70 +1,56 @@
-import 'package:devtools_app/src/ui/colors.dart';
 import 'package:flutter/material.dart';
 
 import '../trees.dart';
+import '../ui/colors.dart';
 import '../utils.dart';
 
 enum PivotType { pivotByMiddle, pivotBySize }
 
-class Treemap extends StatefulWidget {
-  const Treemap({
-    @required this.rootNode,
-    @required this.levelsVisible,
-    @required this.width,
-    @required this.height,
-    @required this.onTap,
-  });
+class NewTreemap extends StatefulWidget {
+  NewTreemap._({
+    this.rootNode,
+    this.nodes = const [],
+    this.levelsVisible,
+    this.onRootChangedCallback,
+  }) : assert(rootNode == null && nodes.isNotEmpty ||
```

Comment:
nit: prefer having two named constructors rather than one constructor where only one of two arguments can be set.

Reply:
I suggested that Peter do it this way. Since it is a widget, all fields have to be final, so the constructor that sets defaults for `rootNode` and `nodes` is private, and the only way to construct a Treemap widget is from `Treemap.fromRoot` or `Treemap.fromNodes`

Reply:
ah I missed that this constructor is private. Actually I don't think this constructor is adding any value as you can just inline it into the other two constructors and then you don't need the assert as the assert is always true and using `this.` the code will be about ~15 lines shorter.

Reply:
Does that mean we should have something like this instead?
```
class Treemap extends StatelessWidget {

  const Treemap.fromRoot({
    @required this.rootNode,
    this.nodes,
    @required this.levelsVisible,
    @required this.outermostLevel,
    @required this.height,
    @required this.onRootChangedCallback,
  });

  const Treemap.fromNodes({
    this.rootNode,
    @required this.nodes,
    @required this.levelsVisible,
    @required this.outermostLevel,
    @required this.height,
    @required this.onRootChangedCallback,
  });

  final TreemapNode rootNode;

  final List<TreemapNode> nodes;
```

Reply:
correct.